### PR TITLE
dev: add `rewrite` flag

### LINKED ISSUE #69570: bazel: support `--rewrite` for datadriven and logic tests
**Describe the problem**

We have various datadriven and logic tests in CRDB that make use of golden datafiles, and support a `--rewrite` flag to update those golden files. Without --rewrite these datafiles are opened in [read-only mode](https://github.com/cockroachdb/datadriven/blob/e89bf9ee1861a9bf55863a1d50a3672a1ceefb18/datadriven.go#L116), so works fine. With `--rewrite` we use the [read-write mode](https://github.com/cockroachdb/datadriven/blob/e89bf9ee1861a9bf55863a1d50a3672a1ceefb18/datadriven.go#L120), which is not permitted over [data dependencies in the sandbox](https://github.com/cockroachdb/cockroach/blob/7057d2fe785535aceaf1db0c418632e305d79dc7/pkg/sql/opt/exec/execbuilder/BUILD.bazel#L57).

Strawman: we could instruct the libraries to write to another path (when run under bazel) and copy those files over those files into source tree. Bazel supports a [`TEST_UNDECLARED_OUTPUTS_DIR`](https://docs.bazel.build/versions/main/test-encyclopedia.html#initial-conditions) envvar that could work as a test-specific scratch space. It doesn't have to be `dev` copying over these files smartly, the tests could do it themselves by checking `bazel.BuiltWithBazel()` provided we include the various testdata directories as part of our `--sandbox_writable_path`. Datadriven for e.g. takes in the path it [reads/writes from](https://github.com/cockroachdb/datadriven/blob/master/datadriven.go#L114); a hacky prototype in some other project:

```go
func cp(t *testing.T, src, dest string) {
	input, err := ioutil.ReadFile(src)
	require.Nil(t, err)
	require.Nil(t, os.MkdirAll(filepath.Dir(dest), 0755))
	if err := ioutil.WriteFile(dest, input, 0644); err != nil {
		t.Fatal(err)
	}
	require.Nil(t, ioutil.WriteFile(dest, input, 0644))
}

func TestDatadriven(t *testing.T) {
	outdir := os.Getenv("TEST_UNDECLARED_OUTPUTS_DIR")
	workspace := os.Getenv("BAZEL_WORKSPACE") // passed in using --test_env=BAZEL_WORKSPACE=<...>, is there a better way?
	datadriven.Walk(t, "testdata", func(t *testing.T, path string) {
		dest := filepath.Join(outdir, path)
		cp(t, path, dest)
		defer func() { cp(t, dest, filepath.Join(repo, "<package name>", path)) }()

		datadriven.RunTest(t, dest, func(t *testing.T, d *datadriven.TestData) string {

```

**To Reproduce**

```sh
$ dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete # read-only mode; passes
$ bazel test --config=devdarwinx86_64 //pkg/sql/opt/exec/execbuilder:execbuilder_test \  # read-write mode; fails
    --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
...
--- FAIL: TestExecBuild (0.29s)
    test_log_scope.go:73: test logs captured to: /var/folders/3j/6fxg9ckn21g5xm0z_06r7gzr0000gn/T/logTestExecBuild048473035
    test_log_scope.go:74: use -show-logs to present logs inline
    --- FAIL: TestExecBuild/local (0.23s)
        --- FAIL: TestExecBuild/local/delete (0.23s)
            logic.go:3127:
                testdata/delete:247: error while processing
            logic.go:3127: open testdata/delete: operation not permitted
    logic.go:3359: -- test log scope end --
FAIL
```

**Expected behavior**

The test to pass.

### PR DESCRIPTION
**Pull request by rickystewart**
Some of our tests support a `-rewrite` flag which updates the `testdata`
on-disk. At a glance this behavior is broken in Bazel, since tests don't
have write access to the `testdata` in the sandbox.

To address this we add a `--rewrite` flag to `dev`. When set,
this flag causes `dev` to `cd` to the subpackage under test in the
workspace before running the test. We also set the environment variable
`YOU_ARE_IN_THE_WORKSPACE` to denote that this happened. We also pass
`-rewrite` to the test.

On the Go side, we check if `YOU_ARE_IN_THE_WORKSPACE` is set and
find `testdata` relative to `PWD` accordingly.

Closes #69570.

Release note: None

### CODE REVIEW

**Comment by rickystewart** [comment]
@irfansharif Not super familiar with the `--rewrite` machinery, can you double-check and make sure this does the trick? Might still need some updates to more of the Go code to find the `YOU_ARE_IN_THE_WORKSPACE` variable and do the appropriate thing.

**Comment by irfansharif** [comment]
```sh
$ dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete --run-in-workspace -- --test_arg -rewrite
bazel test //pkg/sql/opt/exec/execbuilder:execbuilder_test --run_under 'cd /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder && ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
INFO: Analyzed target //pkg/sql/opt/exec/execbuilder:execbuilder_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //pkg/sql/opt/exec/execbuilder:execbuilder_test up-to-date:
  _bazel/bin/pkg/sql/opt/exec/execbuilder/execbuilder_test_/execbuilder_test
INFO: Elapsed time: 0.726s, Critical Path: 0.36s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
//pkg/sql/opt/exec/execbuilder:execbuilder_test                 (cached) PASSED in 0.1s

Executed 0 out of 1 test: 1 test passes.
INFO: Build completed successfully, 1 total action
```

Worked on my mac, but opaquely fails on my gceworker (after rebasing on top of https://github.com/cockroachdb/cockroach/commit/3764d2a09d63cf64c659eacf448dc00ba7744c1b). Ran out of time to investigate, like I'm doing something dumb.

```sh
$ ./dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete --run-in-workspace -- --test_arg -rewrite                         21:45:23
bazel test //pkg/sql/opt/exec/execbuilder:execbuilder_test --run_under 'cd /home/irfansharif/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder &&
 ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
INFO: Build options --define, --run_under, --test_arg, and 2 more have changed, discarding analysis cache.
INFO: Analyzed target //pkg/sql/opt/exec/execbuilder:execbuilder_test (1 packages loaded, 24692 targets configured).
INFO: Found 1 test target...
ERROR: /home/irfansharif/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/BUILD.bazel:49:8: Testing //pkg/sql/opt/exec/execbuilder:execbuilder_tes
t failed: unexpected I/O exception: /artifacts/tmp/_tmp/14c81c8421bd3f7433b36b1dcf0e6b77 (No such file or directory)
Target //pkg/sql/opt/exec/execbuilder:execbuilder_test up-to-date:
  _bazel/bin/pkg/sql/opt/exec/execbuilder/execbuilder_test_/execbuilder_test
INFO: Elapsed time: 1.921s, Critical Path: 0.94s
INFO: 4 processes: 2 internal, 2 linux-sandbox.
FAILED: Build did NOT complete successfully
//pkg/sql/opt/exec/execbuilder:execbuilder_test                       NO STATUS

FAILED: Build did NOT complete successfully
ERROR: exit status 36
```

Instead of `--run-in-workspace`, what do you think of just promoting this to `--rewrite` and also then including `--test_arg -rewrite`? It would help keep the `dev` vocabulary to thinks we already use for the makefile; our equivalent today is `make test ... TESTFLAGS='-rewrite'`. Or are you thinking this flag would be more generally used?

Seems like there's a tiny caveat with using `--run_under`, i.e not being able to use `-rewrite` against multiple test targets at once (something we usually do). It feels a bit unfortunate, and perhaps it's fine -- I'm still coming from the POV that using UNDECLARED_OUTPUT_DIR is not entirely inappropriate for datadriven tests:

https://github.com/irfansharif/solver/blob/11071354d9bd31c19e24f15f57afd8404ea5eee0/datadriven_test.go#L32-L34

https://github.com/irfansharif/solver/blob/11071354d9bd31c19e24f15f57afd8404ea5eee0/internal/testutils/bazel/bazel.go#L33-L37

That approach introduces the (probably more annoying) caveat that all the testdata directories need to be whitelisted. I imagined doing something like `bazel query "rdeps(..., @com_github_cockroachdb_datadriven//:datadriven)` to get the test targets and going from there, but sounds annoying.

---

TL;DR LGTM.

**Review by irfansharif — APPROVED** [review]
```sh
$ dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete --run-in-workspace -- --test_arg -rewrite
bazel test //pkg/sql/opt/exec/execbuilder:execbuilder_test --run_under 'cd /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder && ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
INFO: Analyzed target //pkg/sql/opt/exec/execbuilder:execbuilder_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //pkg/sql/opt/exec/execbuilder:execbuilder_test up-to-date:
  _bazel/bin/pkg/sql/opt/exec/execbuilder/execbuilder_test_/execbuilder_test
INFO: Elapsed time: 0.726s, Critical Path: 0.36s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
//pkg/sql/opt/exec/execbuilder:execbuilder_test                 (cached) PASSED in 0.1s

Executed 0 out of 1 test: 1 test passes.
INFO: Build completed successfully, 1 total action
```

Worked on my mac, but opaquely fails on my gceworker (after rebasing on top of https://github.com/cockroachdb/cockroach/commit/3764d2a09d63cf64c659eacf448dc00ba7744c1b). Ran out of time to investigate, like I'm doing something dumb.

```sh
$ ./dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete --run-in-workspace -- --test_arg -rewrite                         21:45:23
bazel test //pkg/sql/opt/exec/execbuilder:execbuilder_test --run_under 'cd /home/irfansharif/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder &&
 ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
INFO: Build options --define, --run_under, --test_arg, and 2 more have changed, discarding analysis cache.
INFO: Analyzed target //pkg/sql/opt/exec/execbuilder:execbuilder_test (1 packages loaded, 24692 targets configured).
INFO: Found 1 test target...
ERROR: /home/irfansharif/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/BUILD.bazel:49:8: Testing //pkg/sql/opt/exec/execbuilder:execbuilder_tes
t failed: unexpected I/O exception: /artifacts/tmp/_tmp/14c81c8421bd3f7433b36b1dcf0e6b77 (No such file or directory)
Target //pkg/sql/opt/exec/execbuilder:execbuilder_test up-to-date:
  _bazel/bin/pkg/sql/opt/exec/execbuilder/execbuilder_test_/execbuilder_test
INFO: Elapsed time: 1.921s, Critical Path: 0.94s
INFO: 4 processes: 2 internal, 2 linux-sandbox.
FAILED: Build did NOT complete successfully
//pkg/sql/opt/exec/execbuilder:execbuilder_test                       NO STATUS

FAILED: Build did NOT complete successfully
ERROR: exit status 36
```

Instead of `--run-in-workspace`, what do you think of just promoting this to `--rewrite` and also then including `--test_arg -rewrite`? It would help keep the `dev` vocabulary to thinks we already use for the makefile; our equivalent today is `make test ... TESTFLAGS='-rewrite'`. Or are you thinking this flag would be more generally used?

Seems like there's a tiny caveat with using `--run_under`, i.e not being able to use `-rewrite` against multiple test targets at once (something we usually do). It feels a bit unfortunate, and perhaps it's fine -- I'm still coming from the POV that using UNDECLARED_OUTPUT_DIR is not entirely inappropriate for datadriven tests:

https://github.com/irfansharif/solver/blob/11071354d9bd31c19e24f15f57afd8404ea5eee0/datadriven_test.go#L32-L34

https://github.com/irfansharif/solver/blob/11071354d9bd31c19e24f15f57afd8404ea5eee0/internal/testutils/bazel/bazel.go#L33-L37

That approach introduces the (probably more annoying) caveat that all the testdata directories need to be whitelisted. I imagined doing something like `bazel query "rdeps(..., @com_github_cockroachdb_datadriven//:datadriven)` to get the test targets and going from there, but sounds annoying.

---

TL;DR LGTM.

**Comment by rickystewart** [comment]
bors r=irfansharif
